starlark/int.go

```diff
@@ -142,75 +152,166 @@ var (
+}
+func (x Int) Sub(y Int) Int {
+	if x.big != nil || y.big != nil {
+		return maybeSmall(new(big.Int).Sub(x.BigInt(), y.BigInt()))
+	}
+	return maybeBig(x.small - y.small)
+}
+func (x Int) Mul(y Int) Int {
+	if x.big != nil || y.big != nil {
+		return maybeSmall(new(big.Int).Mul(x.BigInt(), y.BigInt()))
+	}
+	return maybeBig(x.small * y.small)
+}
+func (x Int) Or(y Int) Int {
+	if x.big != nil || y.big != nil {
+		return Int{big: new(big.Int).Or(x.BigInt(), y.BigInt())}
+	}
+	return Int{small: x.small | y.small}
+}
+func (x Int) And(y Int) Int {
+	if x.big != nil || y.big != nil {
+		return maybeSmall(new(big.Int).And(x.BigInt(), y.BigInt()))
+	}
+	return Int{small: x.small & y.small}
+}
+func (x Int) Xor(y Int) Int {
+	if x.big != nil || y.big != nil {
+		return maybeSmall(new(big.Int).Xor(x.BigInt(), y.BigInt()))
+	}
+	return Int{small: x.small ^ y.small}
+}
+func (x Int) Not() Int {
+	if x.big != nil {
+		return maybeSmall(new(big.Int).Not(x.big))
+	}
+	//return maybeBig(^x.small)
+	return Int{small: ^x.small}
+}
+func (x Int) Lsh(y uint) Int { return maybeSmall(new(big.Int).Lsh(x.BigInt(), y)) }
+func (x Int) Rsh(y uint) Int { return maybeSmall(new(big.Int).Rsh(x.BigInt(), y)) }
 
 // Precondition: y is nonzero.
 func (x Int) Div(y Int) Int {
 	// http://python-history.blogspot.com/2010/08/why-pythons-integer-division-floors.html
-	var quo, rem big.Int
-	quo.QuoRem(x.bigint, y.bigint, &rem)
-	if (x.bigint.Sign() < 0) != (y.bigint.Sign() < 0) && rem.Sign() != 0 {
-		quo.Sub(&quo, one.bigint)
+	if x.big != nil || y.big != nil {
+		xb, yb := x.BigInt(), y.BigInt()
+
+		var quo, rem big.Int
+		quo.QuoRem(xb, yb, &rem)
+		if (xb.Sign() < 0) != (yb.Sign() < 0) && rem.Sign() != 0 {
+			quo.Sub(&quo, oneBig)
+		}
+		return maybeSmall(&quo)
 	}
-	return Int{&quo}
+	quo := x.small / y.small
+	rem := x.small % y.small
+	if (signum64(x.small) < 0) != (signum64(y.small) < 0) && signum64(rem) != 0 {
+		quo -= 1
+	}
+	return maybeBig(quo)
 }
 
 // Precondition: y is nonzero.
 func (x Int) Mod(y Int) Int {
-	var quo, rem big.Int
-	quo.QuoRem(x.bigint, y.bigint, &rem)
-	if (x.bigint.Sign() < 0) != (y.bigint.Sign() < 0) && rem.Sign() != 0 {
-		rem.Add(&rem, y.bigint)
+	if x.big != nil || y.big != nil {
+		xb, yb := x.BigInt(), y.BigInt()
+
+		var quo, rem big.Int
+		quo.QuoRem(xb, yb, &rem)
+		if (xb.Sign() < 0) != (yb.Sign() < 0) && rem.Sign() != 0 {
+			rem.Add(&rem, yb)
+		}
+		return maybeSmall(&rem)
+	}
+	rem := x.small % y.small
+	if (signum64(x.small) < 0) != (signum64(y.small) < 0) && signum64(rem) != 0 {
+		rem += y.small
+	}
+	return Int{small: rem}
+}
+
+func maybeSmall(x *big.Int) Int {
+	if x.BitLen() < 32 {
+		return Int{small: x.Int64()}
+	}
+	return Int{big: x}
+}
+
+func maybeBig(x int64) Int {
+	if math.MinInt32 <= x && x <= math.MaxInt32 {
+		return Int{small: x}
 	}
-	return Int{&rem}
+	return Int{big: newBig(x)}
 }
 
-func (i Int) rational() *big.Rat { return new(big.Rat).SetInt(i.bigint) }
+func (i Int) rational() *big.Rat {
```

Comment:
Let's inline this; it's called only once. Then we can eliminate the following nil check.

Reply:
Its called twice here: https://github.com/google/starlark-go/blob/master/starlark/value.go#L1160 and https://github.com/google/starlark-go/blob/master/starlark/value.go#L1175

Reply:
Do floating comparisons need to be done with math big?